packages/dds/matrix/src/matrix.ts

```diff
@@ -788,36 +772,43 @@ export class SharedMatrix<T = any>
 				this.rows.applyMsg(msg, local);
 				break;
 			default: {
-				if (contents.type === MatrixOp.changeSetCellPolicy) {
-					this.setCellLwwToFwwPolicySwitchOpSeqNumber = rawMessage.sequenceNumber;
-					return;
-				}
-
 				assert(
 					contents.type === MatrixOp.set,
 					0x021 /* "SharedMatrix message contents have unexpected type!" */,
 				);
 
-				const { row, col, value } = contents;
+				const { row, col, value, fwwMode } = contents;
+				const isPreviousSetCellPolicyModeFWW =
+					this.setCellLwwToFwwPolicySwitchOpSeqNumber > -1;
+				// If this is the first op notifying us of the policy change, then set the policy change seq number.
+				if (this.setCellLwwToFwwPolicySwitchOpSeqNumber === -1 && fwwMode === true) {
+					this.setCellLwwToFwwPolicySwitchOpSeqNumber = rawMessage.sequenceNumber;
+				}
+
 				assert(rawMessage.clientId !== null, "clientId should not be null!!");
 				if (local) {
 					// We are receiving the ACK for a local pending set operation.
 					const { rowHandle, colHandle, localSeq } = localOpMetadata as ISetOpMetadata;
-					if (this.isSetCellOpMadeInFWWPolicy(rawMessage)) {
-						if (this.shouldSetCellBasedOnFWW(rowHandle, colHandle, rawMessage)) {
-							this.cellLastWriteTracker.setCell(rowHandle, colHandle, {
-								seqNum: rawMessage.sequenceNumber,
-								clientId: rawMessage.clientId,
-							});
-						}
-					} else {
-						if (this.isLatestPendingWrite(rowHandle, colHandle, localSeq)) {
-							this.pending.setCell(rowHandle, colHandle, undefined);
-							this.cellLastWriteTracker.setCell(rowHandle, colHandle, {
-								seqNum: rawMessage.sequenceNumber,
-								clientId: rawMessage.clientId,
-							});
-						}
+					const isLatestPendingOp = this.isLatestPendingWrite(
+						rowHandle,
+						colHandle,
+						localSeq,
+					);
+					// If policy is switched and cell should be modified too based on policy, then update the tracker.
+					// If policy is not switched, then also update the tracker in case it is the latest.
+					if (
+						(this.setCellLwwToFwwPolicySwitchOpSeqNumber > -1 &&
+							this.shouldSetCellBasedOnFWW(rowHandle, colHandle, rawMessage)) ||
+						(this.setCellLwwToFwwPolicySwitchOpSeqNumber === -1 && isLatestPendingOp)
```

Comment:
Will isLatestPendingOp check produce correct result? I think it does, but worth double-checking this sequence is correct:

1. This client issued a first op to switch to FWW.
   - All clients will apply it (as op itself always applied, it's sort of the last op treated as LWW).
   - This client will not apply it, as it's local and not the last local change by this client.
2.  There is some other op from another client that is accepted (as it was generated with ref seq # with from # 1).
3. The last op from this client makes it through
   - it's rejected by all clients as it conflicts with # 2.

I think that's fine, as # 2 will overwrite the state for all clients (including this client).
Do we have enough test coverage (including fuzz testing) to force these cases, and ensure that at least we are eventually consistent?

Reply:
Yes, the above sequence works, and we have test also for this.